pyreal/explainers/generic_explainer.py

```diff
@@ -0,0 +1,119 @@
+           x_orig -> x_explain
+        m_transforms (transformer object or list of transformer objects):
+           Transformer(s) needed on x_orig to make predictions on the dataset with model,
+           if different than e_transformers
+           x_orig -> x_model
+        i_transforms (transformer object or list of transformer objects):
+           Transformer(s) needed to make x_orig interpretable
+           x_orig -> x_interpret
+        interpretable_features (Boolean):
+            If True, return explanations using the interpretable feature descriptions instead of
+            default names
+
+    Returns:
+        Explainer:
+            The fitted explainer. Only returned in return_explainer is True
+        DataFrame of shape (n_instances, n_features):
+            The contribution of each feature. Only returned if return_contributions is True
+    """
+    if not return_explanation and not return_explainer:
+        raise ValueError("return_explanation and return_explainer cannot both be false")
+    if explainer is None and (model is None or x_train_orig is None):
+        raise ValueError("You must provide either explainer OR model and corresponding x_train")
+
+    if explainer is None:
+        explainer = Explainer(model, x_train_orig,
+                              feature_descriptions=feature_descriptions,
+                              e_transforms=e_transforms, m_transforms=m_transforms,
+                              i_transforms=i_transforms,
+                              interpretable_features=interpretable_features,
+                              fit_on_init=True)
+    if return_explainer and return_explanation:
+        return explainer, explainer.produce(x_orig)
+    if return_explainer:
+        return explainer
+    if return_explanation:
+        return explainer.produce(x_orig)
+
+
+class Explainer(BaseExplainer):
+    """
+    Unspecified Explainer wrapper.
+
+    A GenericExplainer object assigns a default ML explainer based on the type of x_orig.
+
+    Args:
+        model (string filepath or model object):
+           Filepath to the pickled model to explain, or model object with .predict() function
+        x_train_orig (dataframe of shape (n_instances, x_orig_feature_count)):
+           The training set for the explainer
+        scope (string of either "global" or "local"):
+            Whether the explainer is global or local
+        interpretable_features (Boolean):
+            If True, return explanations using the interpretable feature descriptions instead of
+            default names
+        **kwargs: see base Explainer args
+    """
+
+    def __init__(self, model, x_train_orig, scope="global", interpretable_features=True, **kwargs):
+        self.scope = scope
+        self.interpretable_features = interpretable_features
+        # TODO: implement smart choosing algorithm based on type of x
+        if scope == "global":
```

Comment:
It *might* be more user friendly to - rather than having generic explainers be either local or global - have them capable of doing *both*. In other words, the user can either call `produce()` or `produce(x_orig)`, and in either case they will get an explanation back (global in first case, local in second). This prevents users from even needing to understand the concept of global vs local. 

We can talk about this more - do you have any thoughts?

Reply:
I think this is a great idea, but our current structure requires us to fit the explainer before calling `produce()`, and when we fit the explainer it is already defined as a global or local explainer. There is a workaround since the fitting process is the same for both global and local explainers, but we might need to define a new explainer class that can produce both global and local explanations.